### Titanic dataset from seaborn

In [29]:
# Usual imports
import numpy as np
import pandas as pd
import seaborn as sns

In [30]:
# Load Titanic dataset from seaborn
df = sns.load_dataset('titanic')
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [31]:
# Check for NaN values
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [32]:
# Deck values have lots of NaN values 688/891 = 77.2%
# Drop deck dolumn, because it hard to impute
df = df.drop('deck', axis=1)
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  
0      man        True  Southampton    no  False  
1    woman       False    Cherbourg   yes  False  
2    woman       False  Southampton   yes   True  
3    woman       False  Southampton   yes  False  
4      man        True  Southampton    no   True  
..     ...         ...          ...   ...    ...  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
888  woman       False  Southampton    no  False  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True  

[891 rows x 14 columns]

In [33]:
# Check again missing values
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
embark_town      2
alive            0
alone            0
dtype: int64

In [34]:
# Check age based values by cross table grouping
# We get mean age by sex, class and embark_town
df.groupby(['class', 'embark_town', 'sex']).mean('age')

C:\Users\jtsil\AppData\Local\Temp\ipykernel_25716\2821323188.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(['class', 'embark_town', 'sex']).mean('age')


survived  pclass        age     sibsp     parch  \
class  embark_town sex                                                       
First  Cherbourg   female  0.976744     1.0  36.052632  0.511628  0.302326   
                   male    0.404762     1.0  40.111111  0.238095  0.333333   
       Queenstown  female  1.000000     1.0  33.000000  1.000000  0.000000   
                   male    0.000000     1.0  44.000000  2.000000  0.000000   
       Southampton female  0.958333     1.0  32.704545  0.604167  0.625000   
                   male    0.354430     1.0  41.897188  0.329114  0.253165   
Second Cherbourg   female  1.000000     2.0  19.142857  0.714286  0.571429   
                   male    0.200000     2.0  25.937500  0.500000  0.500000   
       Queenstown  female  1.000000     2.0  30.000000  0.000000  0.000000   
                   male    0.000000     2.0  57.000000  0.000000  0.000000   
       Southampton female  0.910448     2.0  29.719697  0.477612  0.626866   
                   male    0.154639     2.0  30.875889  0.329897  0.195876   
Third  Cherbourg   female  0.652174     3.0  14.062500  0.565217  0.826087   
                   male    0.232558     3.0  25.016800  0.232558  0.139535   
       Queenstown  female  0.727273     3.0  22.850000  0.212121  0.242424   
                   male    0.076923     3.0  28.142857  0.589744  0.128205   
       Southampton female  0.375000     3.0  23.223684  1.238636  1.000000   
                   male    0.128302     3.0  26.574766  0.528302  0.252830   

                                 fare  adult_male     alone  
class  embark_town sex                                       
First  Cherbourg   female  115.640309    0.000000  0.418605  
                   male     93.536707    1.000000  0.547619  
       Queenstown  female   90.000000    0.000000  0.000000  
                   male     90.000000    1.000000  0.000000  
       Southampton female   99.026910    0.000000  0.291667  
                   male     52.949947    0.962025  0.658228  
Second Cherbourg   female   25.268457    0.000000  0.285714  
                   male     25.421250    0.900000  0.400000  
       Queenstown  female   12.350000    0.000000  1.000000  
                   male     12.350000    1.000000  1.000000  
       Southampton female   21.912687    0.000000  0.417910  
                   male     19.232474    0.917526  0.690722  
Third  Cherbourg   female   14.694926    0.000000  0.217391  
                   male      9.352237    0.906977  0.767442  
       Queenstown  female   10.307833    0.000000  0.757576  
                   male     11.924251    0.897436  0.743590  
       Southampton female   18.670077    0.000000  0.340909  
                   male     13.307149    0.924528  0.762264

In [35]:
# Create helper function to fill missing NaN values in age column
# Use age mean from previous cell block
def impute_age(row):
    pclass = row['pclass']
    age = row['age']
    town = row['embark_town']
    sex = row['sex']

    if pd.isnull(age):
        if pclass == 1:
            if town == 'Cherbourg':
                if sex == 'male':
                    return 40
                else:
                    return 36
            elif town == 'Queenstown':
                if sex == 'male':
                    return 44
                else:
                    return 33
            else:
                if sex == 'male':
                    return 41
                else:
                    return 32
        elif pclass == 2:
            if town == 'Cherbourg':
                if sex == 'male':
                    return 26
                else:
                    return 19
            elif town == 'Queenstown':
                if sex == 'male':
                    return 57
                else:
                    return 30
            else:
                if sex == 'male':
                    return 30
                else:
                    return 30
        else:
            if town == 'Cherbourg':
                if sex == 'male':
                    return 25
                else:
                    return 14
            elif town == 'Queenstown':
                if sex == 'male':
                    return 28
                else:
                    return 23
            else:
                if sex == 'male':
                    return 26
                else:
                    return 23
                
    else:
        return age


In [36]:
# Check NaN values in age column
df.tail(15)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
876         0       3    male  20.0      0      0   9.8458        S   Third   
877         0       3    male  19.0      0      0   7.8958        S   Third   
878         0       3    male   NaN      0      0   7.8958        S   Third   
879         1       1  female  56.0      0      1  83.1583        C   First   
880         1       2  female  25.0      0      1  26.0000        S  Second   
881         0       3    male  33.0      0      0   7.8958        S   Third   
882         0       3  female  22.0      0      0  10.5167        S   Third   
883         0       2    male  28.0      0      0  10.5000        S  Second   
884         0       3    male  25.0      0      0   7.0500        S   Third   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  
876    man        True  Southampton    no   True  
877    man        True  Southampton    no   True  
878    man        True  Southampton    no   True  
879  woman       False    Cherbourg   yes  False  
880  woman       False  Southampton   yes  False  
881    man        True  Southampton    no   True  
882  woman       False  Southampton    no   True  
883    man        True  Southampton    no   True  
884    man        True  Southampton    no   True  
885  woman       False   Queenstown    no  False  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
888  woman       False  Southampton    no  False  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True

In [37]:
# Apply helper function to age column
df['age'] = df.apply(impute_age, axis=1)

In [38]:
# Check if NaN values are replaced
df.tail(15)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
876         0       3    male  20.0      0      0   9.8458        S   Third   
877         0       3    male  19.0      0      0   7.8958        S   Third   
878         0       3    male  26.0      0      0   7.8958        S   Third   
879         1       1  female  56.0      0      1  83.1583        C   First   
880         1       2  female  25.0      0      1  26.0000        S  Second   
881         0       3    male  33.0      0      0   7.8958        S   Third   
882         0       3  female  22.0      0      0  10.5167        S   Third   
883         0       2    male  28.0      0      0  10.5000        S  Second   
884         0       3    male  25.0      0      0   7.0500        S   Third   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female  23.0      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  
876    man        True  Southampton    no   True  
877    man        True  Southampton    no   True  
878    man        True  Southampton    no   True  
879  woman       False    Cherbourg   yes  False  
880  woman       False  Southampton   yes  False  
881    man        True  Southampton    no   True  
882  woman       False  Southampton    no   True  
883    man        True  Southampton    no   True  
884    man        True  Southampton    no   True  
885  woman       False   Queenstown    no  False  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
888  woman       False  Southampton    no  False  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True

In [39]:
# Check what NaN values still exists
df.isna().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

In [40]:
# Drop rows with NaN values, since there is only few of them
df = df.dropna()

In [41]:
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female  23.0      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  
0      man        True  Southampton    no  False  
1    woman       False    Cherbourg   yes  False  
2    woman       False  Southampton   yes   True  
3    woman       False  Southampton   yes  False  
4      man        True  Southampton    no   True  
..     ...         ...          ...   ...    ...  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
888  woman       False  Southampton    no  False  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True  

[889 rows x 14 columns]

In [42]:
# Use factorize to automatically assing number for each category
values, names = pd.factorize(df['sex'], sort=False)

In [43]:
# Assing binary category to sex column 
df['sex'] = values

C:\Users\jtsil\AppData\Local\Temp\ipykernel_25716\2440428984.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['sex'] = values


In [44]:
# Check how well common statistical measurements relate to data
df.describe()

survived      pclass         sex         age       sibsp       parch  \
count  889.000000  889.000000  889.000000  889.000000  889.000000  889.000000   
mean     0.382452    2.311586    0.350956   29.211665    0.524184    0.382452   
std      0.486260    0.834700    0.477538   13.271115    1.103705    0.806761   
min      0.000000    1.000000    0.000000    0.420000    0.000000    0.000000   
25%      0.000000    2.000000    0.000000   22.000000    0.000000    0.000000   
50%      0.000000    3.000000    0.000000   27.000000    0.000000    0.000000   
75%      1.000000    3.000000    1.000000   36.000000    1.000000    0.000000   
max      1.000000    3.000000    1.000000   80.000000    8.000000    6.000000   

             fare  
count  889.000000  
mean    32.096681  
std     49.697504  
min      0.000000  
25%      7.895800  
50%     14.454200  
75%     31.000000  
max    512.329200

In [45]:
# NaN values looking good
df.isna().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

In [46]:
# Create mapper for embark_town column to replace ordinal category
c_map = {'Cherbourg': 1, 'Queenstown': 2, 'Southampton': 3}
# Apply mapper for new column
df['Embark_town'] = df['embark_town'].replace(c_map)
df

C:\Users\jtsil\AppData\Local\Temp\ipykernel_25716\1974268816.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Embark_town'] = df['embark_town'].replace(c_map)


survived  pclass  sex   age  sibsp  parch     fare embarked   class  \
0           0       3    0  22.0      1      0   7.2500        S   Third   
1           1       1    1  38.0      1      0  71.2833        C   First   
2           1       3    1  26.0      0      0   7.9250        S   Third   
3           1       1    1  35.0      1      0  53.1000        S   First   
4           0       3    0  35.0      0      0   8.0500        S   Third   
..        ...     ...  ...   ...    ...    ...      ...      ...     ...   
886         0       2    0  27.0      0      0  13.0000        S  Second   
887         1       1    1  19.0      0      0  30.0000        S   First   
888         0       3    1  23.0      1      2  23.4500        S   Third   
889         1       1    0  26.0      0      0  30.0000        C   First   
890         0       3    0  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  Embark_town  
0      man        True  Southampton    no  False            3  
1    woman       False    Cherbourg   yes  False            1  
2    woman       False  Southampton   yes   True            3  
3    woman       False  Southampton   yes  False            3  
4      man        True  Southampton    no   True            3  
..     ...         ...          ...   ...    ...          ...  
886    man        True  Southampton    no   True            3  
887  woman       False  Southampton   yes   True            3  
888  woman       False  Southampton    no  False            3  
889    man        True    Cherbourg   yes   True            1  
890    man        True   Queenstown    no   True            2  

[889 rows x 15 columns]

In [47]:
# Create helper function for age_groups
def age_group(row):
    age = row['age']

    if age < 30:
        return 1
    elif 30 <= age < 50:
        return 2
    else:
        return 3

In [48]:
# Apply helper function for new column
df['age_group'] = df.apply(age_group, axis=1)

C:\Users\jtsil\AppData\Local\Temp\ipykernel_25716\2660450543.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['age_group'] = df.apply(age_group, axis=1)


In [49]:
df

survived  pclass  sex   age  sibsp  parch     fare embarked   class  \
0           0       3    0  22.0      1      0   7.2500        S   Third   
1           1       1    1  38.0      1      0  71.2833        C   First   
2           1       3    1  26.0      0      0   7.9250        S   Third   
3           1       1    1  35.0      1      0  53.1000        S   First   
4           0       3    0  35.0      0      0   8.0500        S   Third   
..        ...     ...  ...   ...    ...    ...      ...      ...     ...   
886         0       2    0  27.0      0      0  13.0000        S  Second   
887         1       1    1  19.0      0      0  30.0000        S   First   
888         0       3    1  23.0      1      2  23.4500        S   Third   
889         1       1    0  26.0      0      0  30.0000        C   First   
890         0       3    0  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  Embark_town  age_group  
0      man        True  Southampton    no  False            3          1  
1    woman       False    Cherbourg   yes  False            1          2  
2    woman       False  Southampton   yes   True            3          1  
3    woman       False  Southampton   yes  False            3          2  
4      man        True  Southampton    no   True            3          2  
..     ...         ...          ...   ...    ...          ...        ...  
886    man        True  Southampton    no   True            3          1  
887  woman       False  Southampton   yes   True            3          1  
888  woman       False  Southampton    no  False            3          1  
889    man        True    Cherbourg   yes   True            1          1  
890    man        True   Queenstown    no   True            2          2  

[889 rows x 16 columns]

In [55]:
# Lets start to check for more deeper info about passengers who survived
df.groupby(['sex'])['survived'].sum().sort_values(ascending=False)

sex
1    231
0    109
Name: survived, dtype: int64

**By sex:**
- 68 %, (231) where female
- 32 %, (109) were male

In [67]:
# Lets start to check for more deeper info about passengers who survived
df.groupby(['embark_town'])['survived'].sum().sort_values(ascending=False)

embark_town
Southampton    217
Cherbourg       93
Queenstown      30
Name: survived, dtype: int64

**By embark town:**
- 64 % embarked from Southampton
- 27 % embarked from Cherbourg
- 9 % embarked from Queenstown

In [56]:
# By sex and age group
df.groupby(['sex', 'age_group'])['survived'].sum().sort_values(ascending=False)

sex  age_group
1    1            130
     2             82
0    1             61
     2             41
1    3             19
0    3              7
Name: survived, dtype: int64

**By sex and age group**
- 56 % females were under 30 years --> 130 person
    - 38 % of all survivors
- 55 % of males were under 30 years --> 61 person
    - 18 % of all survivors
- 92 % females were under 50 years --> 212 person
    - 62 % of all survivors

In [68]:
# By sex and embark town
df.groupby(['sex', 'embark_town'])['survived'].sum().sort_values(ascending=False)

sex  embark_town
1    Southampton    140
0    Southampton     77
1    Cherbourg       64
0    Cherbourg       29
1    Queenstown      27
0    Queenstown       3
Name: survived, dtype: int64

**By sex and embark town**
- 60 % of females embarked from Southampton
- 28 % of females embarked from Cherbourg
- 12 % of females embarked from Queenstown
- 70 % of males embarked from Southampton
- 27 % of males embarked from Cherbourg
- 3 % of males embarked from Queenstown

In [69]:
# By sex, embark town and age group
df.groupby(['sex', 'embark_town', 'age_group'])['survived'].sum().sort_values(ascending=False)

sex  embark_town  age_group
1    Southampton  1            72
                  2            56
0    Southampton  1            40
1    Cherbourg    1            34
0    Southampton  2            32
1    Queenstown   1            24
     Cherbourg    2            23
0    Cherbourg    1            18
1    Southampton  3            12
0    Cherbourg    2             9
1    Cherbourg    3             7
0    Southampton  3             5
1    Queenstown   2             3
0    Queenstown   1             3
     Cherbourg    3             2
     Queenstown   3             0
                  2             0
Name: survived, dtype: int64

**By sex, embark town and age group**
- 31 % were females under 30 who embarked from Southampton
- 56 % were females under 50 who embarked from Southampton
- 17 % were males under 30 who embarked from Southampton

In [70]:
# By sex, embark town, age group and pclass
df.groupby(['sex', 'age_group', 'embark_town', 'pclass'])['survived'].sum().sort_values(ascending=False)

sex  age_group  embark_town  pclass
1    1          Southampton  2         31
                             3         25
     2          Southampton  2         25
     1          Queenstown   3         24
0    1          Southampton  3         24
1    2          Southampton  1         24
                Cherbourg    1         23
0    2          Southampton  1         17
1    1          Southampton  1         16
                Cherbourg    3         15
                             1         12
0    2          Southampton  3         10
     1          Cherbourg    3         10
                Southampton  2          9
     2          Cherbourg    1          9
1    2          Southampton  3          7
     1          Cherbourg    2          7
0    1          Southampton  1          7
1    3          Cherbourg    1          7
0    1          Cherbourg    1          6
1    3          Southampton  1          6
0    2          Southampton  2          5
1    3          Southampton  2          

**By sex, age group, embark town and pclass**
- 24 % were females under 30 years who embarked from Southampton and travelled in 2nd or 3dr class.

### Conclusions:
- Typical survivor was a female age under 30 years who embarked from Southampton and travelled in 2nd or 3rd class (24 %)
- Majority of survivors were females (68 %)
- Majority of survivors were females under 30 years (38 %)